apis/core/v1beta1/function_webhook.go

```diff
@@ -38,4 +188,492 @@ var _ webhook.Defaulter = &Function{}
 // Default implements webhook.Defaulter so a webhook will be registered for the type
 func (r *Function) Default() {
 	functionlog.Info("default", "name", r.Name)
+	if r.Spec.Version == nil || *r.Spec.Version == "" {
+		version := "latest"
+		r.Spec.Version = &version
+	}
+}
+
+// +kubebuilder:webhook:path=/validate-core-openfunction-io-v1beta1-function,mutating=false,failurePolicy=fail,groups=core.openfunction.io,resources=functions,verbs=create;update,versions=v1beta1,name=vfunctions.of.io,sideEffects=None,admissionReviewVersions=v1
+var _ webhook.Validator = &Function{}
+
+func (r *Function) ValidateCreate() error {
+	functionlog.Info("validate create", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateUpdate(old runtime.Object) error {
+	functionlog.Info("validate update", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateDelete() error {
+	functionlog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+func (r *Function) Validate() error {
+	if r.Spec.Image == "" {
+		return field.Required(field.NewPath("spec", "image"), "must be specified")
+	}
+
+	if r.Spec.Build != nil {
+		if r.Spec.ImageCredentials == nil {
+			return field.Required(field.NewPath("spec", "imageCredentials"),
+				"must be specified when `spec.build` is enabled")
+		}
+		if r.Spec.ImageCredentials != nil && r.Spec.ImageCredentials.Name == "" {
+			return field.Required(field.NewPath("spec", "imageCredentials", "name"),
+				"must be specified when `spec.build` is enabled")
+		}
+		if err := r.ValidateBuild(); err != nil {
+			return err
+		}
+	}
+
+	if r.Spec.Serving != nil {
+		if err := r.ValidateServing(); err != nil {
+			return err
+		}
+	}
+
+	if r.Spec.Build == nil && r.Spec.Serving == nil {
+		return field.Required(field.NewPath("spec", "serving"),
+			"must be specified when `spec.build` is not enabled")
+	}
+	return nil
+}
+
+func (r *Function) ValidateBuild() error {
+	if r.Spec.Build.Builder == nil && r.Spec.Build.Dockerfile == nil {
```

Comment:
`Spec.Build.Builder` will never be nil.

https://github.com/OpenFunction/OpenFunction/blob/main/apis/core/v1beta1/function_types.go#L88

Reply:
@tpiperatgod @benjaminhuo  Can we confirm that Spec.Build.Builder will never be nil?

Reply:
If `spec.build.dockerfile` has a value then can `spec.build.builder` be nil, right? @wanjunlei

Reply:
yeah, we might need to support the dockerfile build method